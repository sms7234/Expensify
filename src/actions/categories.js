import database from '../firebase/firebase';

export const setCategories = (categories) => ({
  type:"SET_CATEGORIES",
  categories
});

export const startSetCategories = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`)
      .once('value')
      .then((snapshot) => {
        const categories = [];
        snapshot.forEach((child) => {
          categories.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setCategories(categories));
      });
  };
};

export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
});

export const startAddCategory = (categoryData = {}) => {
  return (dispatch, getState) =>{
    const {
      category= '',
      description= ''
    } = categoryData;
    const categoryItem = {category, description};
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`).push(categoryItem).then((ref)=>{
      dispatch(addCategory({
        id: ref.key,
        ...categoryItem
      }));
    });
  };
};

export const editCategory = (id, updates) => ({
  type: 'EDIT_CATEGORY',
  id,
  updates
});

export const startEditCategory = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).update(updates).then(()=>{
      dispatch(editCategory(id,updates));
    });
  };
};

export const removeCategory = (id) => ({
  type: 'REMOVE_CATEGORY',
  id
});

export const startRemoveCategory = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).remove().then(()=>{
      dispatch(removeCategory(id));
    });
  };
};
