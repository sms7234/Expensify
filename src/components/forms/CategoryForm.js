import React from 'react';

export default class CategoryForm extends React.Component{
  constructor(props){
    super(props);
    this.state={
      category: props.categories ? props.categories.category: '',
      description: props.categories ? props.categories.description:'',
      error:''
    };
  }
  onCategoryChange = (e) => {
    const category = e.target.value;
    if(category.length > 0){
      this.setState(()=> ({category}));
    };
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(()=>({description}))
  };
  onSubmit = (e) => {
    e.preventDefault();
    if(!this.state.category){
      this.setState(()=> ({error: 'Please fill in category'}));
    } else {
      this.setState(()=>({error:''}));
      this.props.onSubmit({
        category: this.state.category,
        description: this.state.description
      })
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error!=='' && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="category name"
          className="text-input"
          value={this.state.category}
          onChange={this.onCategoryChange}
          autoFocus
        />
        <textarea
          placeholder="Add description for your category"
          className="textarea"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <button className="button">
          Save
        </button>
      </form>
    )
  }
}
