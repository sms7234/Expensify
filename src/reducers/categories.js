export default (state = [], action) => {
  switch(action.type){
    case 'ADD_CATEGORY':
      return state.concat(action.category);
    case 'SET_CATEGORIES':
      return action.categories;
    default:
      return state;
  }
};
