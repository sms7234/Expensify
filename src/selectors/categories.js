export default (categories, {category}) => {
  return categories.filter((item) => {
    const textMatch = item.category.toLowerCase().includes(category.toLowerCase());
    return textMatch;
  }).sort((a,b)=>{
    if(a.category < b.category){
      return -1;
    } else {
      return 1;
    }
  });
};
