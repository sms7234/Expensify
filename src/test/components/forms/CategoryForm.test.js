import React from 'react';
import {shallow} from 'enzyme';
import CategoryForm from '../../../components/forms/CategoryForm';
import categories from '../../fixtures/categories';

test('should render Category Form with no data', () => {
  const wrapper = shallow(<CategoryForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render CategoryForm with data', () => {
  const wrapper = shallow(<CategoryForm category={categories[0]}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid submission', () => {
  const wrapper = shallow(<CategoryForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set category value on change', () => {
  const value = "new category";
  const wrapper= shallow(<CategoryForm />);
  wrapper.find('input').simulate('change', {
    target: {value}
  });
  expect(wrapper.state('category')).toBe(value);
});

test('should set category value on change', () => {
  const value = "new description";
  const wrapper= shallow(<CategoryForm />);
  wrapper.find('textarea').simulate('change', {
    target: {value}
  });
  expect(wrapper.state('description')).toBe(value);
});
