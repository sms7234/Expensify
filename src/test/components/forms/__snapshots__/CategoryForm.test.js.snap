// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Category Form with no data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="category name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your category"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render CategoryForm with data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="category name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your category"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render error for invalid submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="category name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your category"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render error for invalid submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please fill in category
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="category name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your category"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;
